### ISSUE TITLE
[RFC] Consuming wix-style-react in SSR projects

### ISSUE THREAD

**Issue by mykas** [wix/wix-style-react#5048]
# Consuming wix-style-react in SSR projects

## The Problem

Currently our users that use SSR in their project are kind of ok with the solution we provide for them to make wix-style-react work for their project. (Our FAQ explains what kind of parameter needs to be enabled in yoshi config to export stylable css and attach a hook for imports to work).

BUT, now that we are planning to migrate to named-imports in order to support ES modules infrastructure - our "quick-hacks" from docs are not enough anymore.

#### 1. CSS-Modules (sass)

`Server` renders an app to generate html with classnames - it will pick commonjs modules from `wix-style-react/dist/src` and the reason for this is because server is NODE based. 

```
<div classname="_-node_modules-wix-style-react-dist-src-Box-Box__root__3wDCc">
```

`Client` assets are beeing created during webpack build and added to html as external stylesheet. During the webpack build it will use ES-modules infrastructure.

```
// app.css
_-node_modules-wix-style-react-dist-es-src-Box-Box__root__bjtF6
```

> Notice we have two problems here:
> 1. CommonJs: `dist/src`, ESmodules: `dist/es/src`
> 2. CommonJs: `root__3wDCc`, ESmodules: `root__bjtF6`

```error
Warning: Prop `className` did not match. Server: "_-node_modules-wix-style-react-dist-src-Box-Box__root__3wDCc _-node_modules-wix-style-react-dist-src-Box-Box__horizontal__VpebB _-node_modules-wix-style-react-dist-src-Box-Box__middle__KW85s" Client: "_-node_modules-wix-style-react-dist-es-src-Box-Box__root__bjtF6 _-node_modules-wix-style-react-dist-es-src-Box-Box__horizontal__2lMkn _-node_modules-wix-style-react-dist-es-src-Box-Box__middle__3WiyU"
```

#### 2. Stylable

Same as with css-modules, but stylable does not generate full path in the classname. The only problem that hash is dependend on stylesheet path.

Server:
```
<div classname="WSRAvatar1045097907--placeholder">
```

Client:
```
// app.stylable.css
WSRAvatar2846992615--placeholder
```

> CommonJs: `1045097907`, ESmodules: `2846992615`

### The output

1. Webpage will have no styles applied, only html because of classname mismatch.
2. React will throw an error in the console.

## Possible Solutions

We can't really fix this issue for stylable files from our side or ask yoshi to fix this as stylable `.st.css` transpilation is done through `StylableWebpackPlugin`. I've wrote to Tom Raviv from stylable and he was interested in fixing this by adding a patching mechanism to unify the classnames generation.

Solving css-modules issues requires more knowledge. My first thought was contacting yoshi team to suggest some kind of unified solution for everybody, but Ran draw a line by saying this:

>Yoshi does not support SSR - We're willing to add options like the one proposed here if you'll contribute them, but we >don't have the capacity to solve SSR related problems right now. If we'll support SSR, which will probably be a part of >yoshi's future, it will include, testing, docs, template and solutions to every bug/feature request you'll experience :)

Which means if we want to support SSR for our users projects with named-improts - we need to develop some kind of solution from our side. 

## Possible Solution (stylable)

Here is what Tom Raviv wrote:

```
you’ll add a param to the stc (@stylable/cli) command you’re running, something 
like --use-source-namespace, then in the *.st.css files that will be copied to 
your dist, a comment will be included, something like:
/* GENERATED: @st-use-source-namespace ../../../src/components/dropdown/dropdown.st.css */

then, once anyone tries to transform those st.css files it uses the 
namespace from the src directory, and not the cjs or es
```

## Possible Solution (CSS-Module) 1

Most users that do SSR in Wix (even though yoshi don't provide any support for it) are using the yoshi-runtime legacy package, you can see that it is not part of their monorepo and yoshi team does not manages it - https://github.com/wix-private/fed-infra/blob/3ea4813a2c12f2d6db284e7768771e6a9b9f70b9/yoshi-runtime/index.js#L24

There is way to use this hook in way that will solve the problem for our users:
```javascript
wixCssModulesRequireHook( {
  generateScopedName: (name, filepath) => {
    if (filepath.indexOf('/node_modules/wix-style-react') > -1) {
      filepath = filepath.replace('dist/src', 'dist/es/src');
    }

    const hashPrefix = require(path.resolve('package.json')).name;

    const generate = genericNames(cssModulesPattren(), {
      context: `${process.cwd()}/src`,
      hashPrefix
    });
    return generate(name, filepath);
  }
});
```

### Pros

1. No big effort from us. Just a doc explaining it.

### Cons

2. Kinda hacky and if users would want to run the same hook for their css-modules it get can messy for them.

## Possible Solution (CSS-Module) 2

Just like other UI libraries we might want to consider providing SSR tooling from our side. For example Material-UI does provide hooks to attach in server that solves the same problem.

```javascript
import express from 'express';
import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { ServerStyleSheets, ThemeProvider } from '@material-ui/core/styles';
import App from './App';
import theme from './theme';

function handleRender(req, res) {
  const sheets = new ServerStyleSheets();

  // Render the component to a string.
  const html = ReactDOMServer.renderToString(
    sheets.collect(
      <ThemeProvider theme={theme}>
        <App />
      </ThemeProvider>,
    ),
  );

  // Grab the CSS from the sheets.
  const css = sheets.toString();

  // Send the rendered page back to the client.
  res.send(renderFullPage(html, css));
}
```

### Pros

- SSR Official support from library side.
- Easy to use

### Cons

- Requires effort. Making sure that it does not collide with current users usage of: `wixCssModulesRequireHook`.
- Our library becomes not just component library but also tools providing library.

## Possible Solution (CSS-Module) 3

There is way to patch our components in `dist/sc` to import sass stylesheets not from `./Component.sass` but actually from `../../es/src/Component/Component.sass`. This fixes the problem of classnames mismatch.

### Pros

- No effort from user side.
- Works well with `wixCssModulesRequireHook`.

### Cons

- Its hacking our output, but we already do anyway.
- Might have some unexpected side effects. Weird debugging.
- Requires effort from our side. (Changinn Artems script that already changes js imports)

**Comment by mykas** [comment]
cc @nitayneeman @shlomitc @itaytay

**Comment by mykas** [comment]
Just want to add something here.

@shlomitc suggest to try this library: https://github.com/standard-things/esm

It suppose to make server import like client, but it I tried this on wix-style-react-mobile project and it failed to load 🤷‍♂ Issues like: 
```error
{ Error: Cannot find module './components/App'
```

**Comment by mykas** [comment]
So I've managed to solve SSR issues for CSS-Modules with solution 3. 

Stylable already released a patching mechanism to fix the same problem for stylable. Together with Tom Raviv we've tested their solution in wix-style-react-mobile and it works! Now whats left is to implement this in wix-ui-core and wix-style-react major versions. I am gonna prepare a migration plan for this with possible drawbacks etc :)

**Comment by mykas** [comment]
## Plan to patch stylable stylesheets

Ok, so why do we even need one? Well, because we might need a breaking change version for `wix-ui-core`. 

If we just patch current wix-ui-core version  some users will start getting these warnings in their console:
```
(client) Duplicate module namespace: EllipsedTooltip2926345178 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/hocs/EllipsedTooltip/EllipsedTooltip.st.css
(client) Duplicate module namespace: Focusable3217384821 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/hocs/Focusable/Focusable.st.css
(client) Duplicate module namespace: Loader704432813 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/components/popover-next/components/Loader.st.css
```
The warning basically tells that two same stylable stylesheets has been loaded on to the web. I am not sure yet about the consequences of this, but Tom from stylable said that this is not good. The reason for this is simple: some components are hardcoded to import from  `dist/src` or `dist/src/es`. 

It seems that even we even had those still in wix-style-react. Look at this PR:  https://github.com/wix/wix-style-react/pull/5149

With mentioned PR I was able to fix the remaining issues for wix-style-react so there wouldn't be a problem for us, but I bet we might see those in `wix-ui-tpa` or even `wix-ui-adi`.

## The Solution

We do a breaking version change for wix-ui-core and patch with stylable solution for ssr. Migrate wix-style-react 7 to use that version and tell @jonathanadler to double check his code if he wants ssr support and migrate to that new wix-ui-core version too. 

@nitayneeman @shlomitc (if you see this) - tell me what do you think about it?

**Comment by mykas** [comment]
I just want to clarify that I have concerns regarding making this stylable patching requiring new wix-ui-core version. 

I dont know if you saw Ran's email regarding stylable 3 available through opt in parameter, but that might require one more breaking change in wix-ui-core. I am not sure about that but it is possible.

**Comment by mykas** [comment]
**tl;dr** 
So, I've talked with Tom Raviv from stylable again and it seems we are kind of safe with those warnings happening for our users that will use previous minor version of wix-style-react. So I am proceeding with patching wix-ui-core without major version.

**long;story**

Just like I mentioned in the previous comment - we have a problem with some of our code in wix-style-react that are having hardcoded `wix-ui-core` imports like these here: https://github.com/wix/wix-style-react/pull/5149

This problem would start to happen after we patch wix-ui-core to unify classnames for SSR cases. Basically these kind of warnings we will get:
```
(client) Duplicate module namespace: EllipsedTooltip2926345178 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/hocs/EllipsedTooltip/EllipsedTooltip.st.css
(client) Duplicate module namespace: Focusable3217384821 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/hocs/Focusable/Focusable.st.css
(client) Duplicate module namespace: Loader704432813 from /Users/tomr/projects/core3/temp/fed/wix-style-react-mobile/node_modules/wix-ui-core/dist/es/src/components/popover-next/components/Loader.st.css
```

What I found out about those errors is that they are just telling us that duplicate classnames are generated in stylable file because we have different javascript files importing same stylesheets from different paths like: `dist/src/` and `dist/es/src`. 

The big question i rised to Tom Ravi is: does it have other kind of side effects apart from duplicate classnames in the same stylesheet? And he said if those stylesheets are the same  - there shouldn't be a problem.

Why do I rise this question if we fixed those issues already in wix-style-react?

Because some our users fixating version of wix-style-react where those hardcoded paths are still there.

There is a way to avoid this by having a major version of wix-ui-core, but I don't think that there is meaningfull reason to do so because those warnings are just warnings and it would happen only for some older versions of version 7.

So I am proceeding with patching wix-ui-core without major version.

**Comment by nitayneeman** [comment]
>Kinda hacky and if users would want to run the same hook for their css-modules it get can messy for them.

Regarding solution 1 - do we have a concrete example of such a mess?

>Might have some unexpected side effects. Weird debugging.

Regarding solution 3 - sounds like a concern, can you elaborate more?

Regarding solution 2 - this is an interesting approach, however I don't think it's supposed to be part of wix-style-react but rather an SSR utility library providing this tooling for component libraries. Honestly this responsibility might be related to `wix-ui` or maybe `node-platform`.

In the short-term I would think to go with solutions 1 or 3 once we understand the consequences ahead. Personally I would bet on 3 because 1 makes our consumers deal with the internals of the `dist` architecture. Solution 2 seems to be a good way in the long-term but here we need to refine the solution and create a work plan.

**Comment by nitayneeman** [comment]
> Just want to add something here.
> 
> @shlomitc suggest to try this library: https://github.com/standard-things/esm
> 
> It suppose to make server import like client, but it I tried this on wix-style-react-mobile project and it failed to load 🤷‍♂ Issues like:
> 
> ```
> { Error: Cannot find module './components/App'
> ```

Not sure exactly how `esm` is supposed to help to solve the stylesheet mismatches. 

The point is **one** of the following: (?)

1. Replace `dist/es` and `dist/src` in favor of the `esm` output and then the bundle won't have stylesheet mismatches originally?
2. Load the bundle using `esm` inside the consuming application?

**Comment by mykas** [comment]
Hey @nitayneeman , thanks for the comments! Just wanted to say that I've already implemented Solution 3 so its done, but I will answer your questions.

> Regarding solution 1 - do we have a concrete example of such a mess?

So currently our users are using this `wixCssModulesRequireHook` to make their css-modules work in ssr, but if we will provide our own solution - they or we will need to figure out how to make it work both for their css-modules and ours. Maybe its easy maybe not, but it can get messy.

> Regarding solution 3 - sounds like a concern, can you elaborate more?

First I thought that the solution was kind of hacky, but then I saw that other tools does the same. Stylable did very similar solution too. So no issues.

And about the @shlomitc suggested solution I am not sure how it solves the problem, but it did not work for me out of the box to really see the benefit.

**Comment by mykas** [comment]
Closing, as this is already solved with Solution 3.
